### PR TITLE
Shared: add in/out barriers with flow state

### PR DESCRIPTION
**Pull request by asgerf**
Adds overloads of `isBarrierIn` and `isBarrierOut` that take a FlowState.
```codeql
predicate isBarrierIn(Node node, FlowState state)
predicate isBarrierOut(Node node, FlowState state)
```

The most immediate use-case for these is to be able to implement these as calls to `isSource` or `isSink`, without losing the flow state:
```codeql
predicate isBarrierIn(Node node, FlowState state) { isSource(node, state) }
predicate isBarrierOut(Node node, FlowState state) { isSink(node, state) }
```

This is needed for the JS data flow migration. Since this isn't merged yet, I haven't added the use sites in this PR, though I have verified that it works in the JS queries where I have used it.

To get a bit more validation, I also ran a C++ experiment with [this commit](https://github.com/github/codeql/commit/313ac43314d8434a9a3d99d462cd39dc3e157902) resulting in [this evaluation](https://github.com/github/codeql-dca-main/tree/data/asgerf/use-inout-barriers-with-flow-state/reports) which seems to look fine.

### CODE REVIEW

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -265,6 +303,7 @@ module MakeImpl<InputSig Lang> {
     }
 
     /** Provides the relevant barriers for a step from `node1` to `node2`. */
+    bindingset[node1, node2]
     pragma[inline]
```
Comment (hvitved) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:307]: Can be removed now.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 2]
```diff
@@ -273,6 +312,17 @@ module MakeImpl<InputSig Lang> {
       not fullBarrier(node2)
     }
 
+    /** Provides the relevant barriers for a step from `node1,state1` to `node2,state2`, including stateless barriers for `node1` to `node2`. */
+    bindingset[node1, state1, node2, state2]
+    pragma[inline]
```
Comment (hvitved) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:317]: Not needed.

**Review by hvitved — CHANGES_REQUESTED** [review]
Looks reasonable to me.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll (hunk 2 above)
Comment (asgerf) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:317]: You mean you want `pragma[inline]` removed because it is implied by `bindingset`?

Comment (hvitved) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:317]: Yes.

Comment (asgerf) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll]: Done

**Review by hvitved — DISMISSED** [review]

**Review by aschackmull — CHANGES_REQUESTED** [review]
This only checks the new barriers on the state-transitioning steps. It also needs to be checked for a lot of other case (state-preserving local steps, jump steps, read steps, store steps, and call edges).

**Comment by asgerf** [comment]
Thanks for catching this @aschackmull.

As discussed offline, I've put check at the `PathNode` level since these barriers are unlikely to affect pruning and so aren't that important to check in `MkStage`. However, the checks weren't quite as simple as initially discussed because `subpaths` and `PartialPathNode` needed to be updated as well.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 3]
```diff
@@ -3888,6 +3903,7 @@ module MakeImpl<InputSig Lang> {
     ) {
       exists(ArgNodeEx arg, ArgumentPosition apos |
         pathNode(mid, arg, state, cc, _, t, ap, _) and
+        not outBarrier(arg, state) and
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:3906]: s/out/in/, right?

**Review by aschackmull — APPROVED** [review]
LGTM

**Comment by asgerf** [comment]
DCA runs look peaceful so I'll go ahead and merge

**Pull request merged**